WalletWasabi.Tests/UnitTests/WabiSabi/Models/MultipartyTransactionTests.cs

```diff
@@ -0,0 +1,327 @@
+			Assert.True(signed.Inputs.All(x => x.HasWitScript()));
+		}
+
+		[Fact]
+		public void AddWithOptimize()
+		{
+			var alice = WabiSabiFactory.CreateAlice();
+
+			var state = new Construction(DefaultParameters).AddInput(alice.Coins.First());
+
+			var script = BitcoinFactory.CreateScript();
+			var bob = new TxOut(alice.Coins.First().Amount/2, script);
+			var withOutput = state.AddOutput(bob);
+			var duplicateOutputNoFee = withOutput.AddOutput(bob).Finalize();
+
+			var tx2 = duplicateOutputNoFee.CreateUnsignedTransaction();
+			var output = Assert.Single(tx2.Outputs);
+			Assert.Equal(script, output.ScriptPubKey);
+			Assert.Equal(alice.Coins.First().Amount, output.Value);
+		}
+
+		[Fact]
+		public void WitnessValidation()
+		{
+			using Key key1 = new();
+			using Key key2 = new();
+
+			var alice1 = WabiSabiFactory.CreateAlice(key: key1);
+			var alice2 = WabiSabiFactory.CreateAlice(key: key2);
+
+			var alice1Coin = alice1.Coins.First();
+			var alice2Coin = alice2.Coins.First();
+
+			var state = new Construction(DefaultParameters).AddInput(alice1Coin).AddInput(alice2Coin);
+
+			// address reuse bad
+			var bob1 = new TxOut(Money.Coins(1), alice1Coin.ScriptPubKey);
+			var withOutput = state.AddOutput(bob1);
+
+			var bob2 = new TxOut(Money.Coins(1), alice2Coin.ScriptPubKey);
+			var noFeeTx = withOutput.AddOutput(bob2).Finalize();
+
+			var tx = noFeeTx.CreateUnsignedTransaction();
+
+			var alice1Tx = tx.Clone();
+			alice1Tx.Sign(key1.GetBitcoinSecret(Network.Main), alice1Coin);
+			var alice2Tx = tx.Clone();
+			alice2Tx.Sign(key2.GetBitcoinSecret(Network.Main), alice2Coin);
+
+			// Only accept valid witnesses
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(1, alice1Tx.Inputs[0].WitScript));
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(1, alice1Tx.Inputs[1].WitScript));
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(0, alice1Tx.Inputs[1].WitScript));
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(0, alice2Tx.Inputs[0].WitScript));
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(0, alice2Tx.Inputs[1].WitScript));
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature, () => noFeeTx.AddWitness(1, alice2Tx.Inputs[0].WitScript));
+
+			// Add Alice 1's signature
+			var alice1Sig = noFeeTx.AddWitness(0, alice1Tx.Inputs[0].WitScript);
+			Assert.False(alice1Sig.IsFullySigned);
+
+			// Witness can only be accepted once per input
+			ThrowsProtocolException(WabiSabiProtocolErrorCode.WitnessAlreadyProvided, () => alice1Sig.AddWitness(0, alice1Tx.Inputs[0].WitScript));
+		}
+
+		[Fact]
+		public void FeeRateValidation()
+		{
+			var feeRate = new FeeRate(new Money(1000L));
+
+			using Key key1 = new();
```

Comment:
I would not mix: `var variable = new Type(...)` with target-type new expressions (`Type variable = new(...)`) if possible. AFAIK WW developers are free to choose what they like better but it's probably better to be consistent at least at file level.

Reply:
Hmm, I cargo culted that from prior tests under the assumption it was the preferred style, and TBQH I still don't understand how that's different than `var key1 = new Key()`... I'm going to defer to my adult supervisors @lontivero & @molnard so that I know for the future

Reply:
`Type variable = new(...)` is a C# 9 feature. So it is relatively new. I like this syntax personally. I personally don't like `var`s too much because then there is much higher chance that I read the code incorrectly - but that's not shared opinion here. 

`using var key1 = new Key()` and `using Key key1 = new()` are equivalent. Good reading here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new

Reply:
Oh I thought `using` was some special construct tied to the target typed new.

Anyway, I refuse to make a decision myself, I don't have enough brain cells to form a preference at this point, I can barely remember if C# is one of those languages that needs trailing `;`s or not... Lucas & David, one of you should tell me which one to use and that's what I will use from now on ;-)

Reply:
No, `using` simply means "call dispose" once the enclosing code block ends.

Reply:
`var` was not introduced in order to give us all the power of strong-typed language without the syntactic cost inherited from Java, what allowed us to evolve from imperative style to declarative style and made possible to have things like lambda expressions, anonymous types, LINQ and the rest of functional programming features. The fact that it came with a shorter syntax is just a plus but it was not a goal by itself.  

In this very specific case both `var x = new Type()` and `Type x = new ()` are equivalent. However, the new target-typed `new` expressions feature was introduced to eliminate redundancy in the syntax, specially there where duplication was required explicitly, for example _field definition_  while `var` is for type inference so, if well both achieve the same they are conceptually very different. 

Use _target-typed `new` expressions_ wherever redundancy can be avoided and `var` everywhere else.

I will write about this is the discussions board because IMO these new c# 9 feature will create problems.